osmoutils/ibc.go

```diff
@@ -2,24 +2,29 @@ package osmoutils
 
 import (
 	"encoding/json"
-
+	sdk "github.com/cosmos/cosmos-sdk/types"
 	transfertypes "github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
 	channeltypes "github.com/cosmos/ibc-go/v4/modules/core/04-channel/types"
 	ibcexported "github.com/cosmos/ibc-go/v4/modules/core/exported"
 )
 
-// NewStringErrorAcknowledgement returns a new instance of Acknowledgement using an Acknowledgement_Error
-// type in the Response field. These errors differ from the IBC errors in that we use custom error strings.
-// NOTE: Acknowledgements are written into state and thus, changes made to error strings included in packet acknowledgements
-// risk an app hash divergence when nodes in a network are running different patch versions of software.
-func NewStringErrorAcknowledgement(err string) channeltypes.Acknowledgement {
-	// ToDo: Do we want to do this or do we want to use the IBC errors and emit the string?
-
-	return channeltypes.Acknowledgement{
-		Response: &channeltypes.Acknowledgement_Error{
-			Error: err,
-		},
+// NewEmitErrorAcknowledgement creates a new error acknowledgement after having emitted an event with the
+// details of the error.
+func NewEmitErrorAcknowledgement(ctx sdk.Context, err error, errorContexts ...string) channeltypes.Acknowledgement {
+	attributes := make([]sdk.Attribute, len(errorContexts)+1)
+	attributes[0] = sdk.NewAttribute("error", err.Error())
+	for i, s := range errorContexts {
+		attributes[i+1] = sdk.NewAttribute("error-context", s)
 	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			"ibc-acknowledgement-error",
+			attributes...,
+		),
+	})
+
+	return channeltypes.NewErrorAcknowledgement(err)
```

Comment:
cref https://github.com/cosmos/ibc-go/blob/ecb845d5e43f53decf48f8ed88c7847a9a4375cb/modules/core/04-channel/types/acknowledgement.go#L32

This makes sense to me!